Firestore/Swift/Source/Codable/FirestoreDecoder.swift

```diff
@@ -0,0 +1,1189 @@
+        defer { self.decoder.codingPath.removeLast() }
+        
+        guard let decoded = try self.decoder.unbox(self.container[self.currentIndex], as: Double.self) else {
+            throw DecodingError.valueNotFound(type, DecodingError.Context(codingPath: self.decoder.codingPath + [_FirestoreKey(index: self.currentIndex)], debugDescription: "Expected \(type) but found null instead."))
+        }
+        
+        self.currentIndex += 1
+        return decoded
+    }
+    
+    public mutating func decode(_ type: String.Type) throws -> String {
+        guard !self.isAtEnd else {
+            throw DecodingError.valueNotFound(type, DecodingError.Context(codingPath: self.decoder.codingPath + [_FirestoreKey(index: self.currentIndex)], debugDescription: "Unkeyed container is at end."))
+        }
+        
+        self.decoder.codingPath.append(_FirestoreKey(index: self.currentIndex))
+        defer { self.decoder.codingPath.removeLast() }
+        
+        guard let decoded = try self.decoder.unbox(self.container[self.currentIndex], as: String.self) else {
+            throw DecodingError.valueNotFound(type, DecodingError.Context(codingPath: self.decoder.codingPath + [_FirestoreKey(index: self.currentIndex)], debugDescription: "Expected \(type) but found null instead."))
+        }
+        
+        self.currentIndex += 1
+        return decoded
+    }
+    
+    public mutating func decode<T : Decodable>(_ type: T.Type) throws -> T {
+        guard !self.isAtEnd else {
+            throw DecodingError.valueNotFound(type, DecodingError.Context(codingPath: self.decoder.codingPath + [_FirestoreKey(index: self.currentIndex)], debugDescription: "Unkeyed container is at end."))
+        }
+        
+        self.decoder.codingPath.append(_FirestoreKey(index: self.currentIndex))
+        defer { self.decoder.codingPath.removeLast() }
+        
+        guard let decoded = try self.decoder.unbox(self.container[self.currentIndex], as: T.self) else {
+            throw DecodingError.valueNotFound(type, DecodingError.Context(codingPath: self.decoder.codingPath + [_FirestoreKey(index: self.currentIndex)], debugDescription: "Expected \(type) but found null instead."))
+        }
+        
+        self.currentIndex += 1
+        return decoded
+    }
+    
+    public mutating func nestedContainer<NestedKey>(keyedBy type: NestedKey.Type) throws -> KeyedDecodingContainer<NestedKey> {
+        self.decoder.codingPath.append(_FirestoreKey(index: self.currentIndex))
+        defer { self.decoder.codingPath.removeLast() }
+        
+        guard !self.isAtEnd else {
+            throw DecodingError.valueNotFound(KeyedDecodingContainer<NestedKey>.self,
+                                              DecodingError.Context(codingPath: self.codingPath,
+                                                                    debugDescription: "Cannot get nested keyed container -- unkeyed container is at end."))
```

Comment:
Despite the different error message, this could probably also use `expectNotAtEnd`. I'm not sure there's any special value in giving a slightly better error message for cases that seem extremely unlikely.

Reply:
Done